src/Shared/CommunicationsUtilities.cs

```diff
@@ -80,8 +81,56 @@ internal enum HandshakeOptions
         /// ARM64 process.
         /// </summary>
         Arm64 = 128,
+
+        /// <summary>
+        /// Using a long-running sidecar TaskHost process to reduce startup overhead and reuse in-memory caches.
+        /// </summary>
+        SidecarTaskHost = 256,
+    }
+
+    /// <summary>
+    /// Status codes for the handshake process.
+    /// It aggregates return values across several functions so we use an aggregate instead of a separate class for each method.
+    /// </summary>
+    internal enum HandshakeStatus
+    {
+        Success = 0,
+        // Other node resurned different value than we expected.
+        // This can happen either by attempting to connect to a wrong node type. (e.g. transient TaskHost trying to connect to a long-running TaskHost)
+        // Or by trying to connect to a node that has a different MSBuild version.
+        VersionMismatch = 1,
+        // TryReadInt -> Abort due to old MSBuild version
+        OldMSBuild = 2,
+        Timeout = 3,
+        UnexpectedEndOfStream = 4,
+        EndiannessMismatch = 5,
+    }
+
+    /// <summary>
+    /// An aggregate class for passing around results of a handshake and adjacent information.
+    /// There used to be a several functions around TryReadIntForHandshake and node communication in general that used
+    /// to throw exceptions to indicate failure. This class is here to replace the exceptions and allow for a different way of signallign the result.
```

Comment:
nit: I don't like elaborations on history of code, why not just describe the current state.

Though I appreciate that you did this refactor.